apiserver/params/cloud.go

```diff
@@ -114,3 +114,67 @@ type CloudSpecResult struct {
 type CloudSpecResults struct {
 	Results []CloudSpecResult `json:"results,omitempty"`
 }
+
+// CloudCredentialRequestParam defines a credential in terms of its cloud and name.
+// It is used to request detailed content for the credential stored on the controller.
+type CloudCredentialRequestParam struct {
+	CloudName      string `json:"cloud-name"`
+	CredentialName string `json:"credential-name"`
+}
+
+func (p CloudCredentialRequestParam) IsEmpty() bool {
+	return p.CloudName == "" && p.CredentialName == ""
+}
+
+// CredentialContentRequestParam defines an input required to make a valid call
+// to get credentials content stored on the controller.
+type CredentialContentRequestParam struct {
+	Credentials    []CloudCredentialRequestParam `json:"credentials,omitempty"`
+	IncludeSecrets bool                          `json:"include-secrets"`
+}
+
+// CloudCredential contains a cloud credential content.
+type CredentialContent struct {
```

Comment:
Would be nice to add to the exiting CloudCredential struct instead of creating a new, similar one.
The the result for the controller credentials api call were a map[cloudname]CloudCredential, then the extra fields in the CloudCredential struct would be name and optional, omitempty secrets. Maybe just leave a TODO. But if we were to use a map in the controller credentials result, that would make the TODO easier to do later

Reply:
I kind of disagree - re-using existing struct means that from some calls it will come back semi-anaemic. I'd prefer to return fully populated objects.. I'll think on it but atm am reluctant to change existing...

Reply:
In addition, existing CloudCredential returns a collection of names of hidden attributes (secrets), whereas I need a collection of names and values of these secrets when they are requested.. So, the details returned are different already...